tool/tbot/main.go

```diff
@@ -0,0 +1,607 @@
+
+	return nil
+}
+
+func onInit(botConfig *config.BotConfig) error {
+	return trace.NotImplemented("init not yet implemented")
+}
+
+func onWatch(botConfig *config.BotConfig) error {
+	return trace.NotImplemented("watch not yet implemented")
+}
+
+func onStart(botConfig *config.BotConfig) error {
+	// Start by loading the bot's primary destination.
+	dest, err := botConfig.Storage.GetDestination()
+	if err != nil {
+		return trace.WrapWithMessage(err, "could not read bot storage destination from config")
+	}
+
+	addr, err := utils.ParseAddr(botConfig.AuthServer)
+	if err != nil {
+		return trace.WrapWithMessage(err, "invalid auth server address %+v", botConfig.AuthServer)
+	}
+
+	var authClient *auth.Client
+
+	// First, attempt to load an identity from the given destination
+	ident, err := identity.LoadIdentity(dest)
+	if err == nil {
+		log.Infof("successfully loaded identity %+v", ident)
+
+		// TODO: we should cache the token; if --token is provided but
+		// different, assume the user is attempting to start with a new
+		// identity
+		if botConfig.Onboarding != nil {
+			log.Warn("note: onboard config ignored as identity was loaded from persistent storage")
+		}
+
+		authClient, err = authenticatedUserClientFromIdentity(ident, botConfig.AuthServer)
+		if err != nil {
+			return trace.Wrap(err)
+		}
+	} else {
+		// If the identity can't be loaded, assume we're starting fresh and
+		// need to generate our initial identity from a token
+
+		// TODO: validate that errors from LoadIdentity are sanely typed; we
+		// actually only want to ignore NotFound errors
+
+		onboarding := botConfig.Onboarding
+		if onboarding == nil {
+			return trace.BadParameter("onboarding config required on first start via CLI or YAML")
+		}
+
+		// If no token is present, we can't continue.
+		if onboarding.Token == "" {
+			return trace.Errorf("unable to start: no identity could be loaded and no token present")
+		}
+
+		tlsPrivateKey, sshPublicKey, tlsPublicKey, err := generateKeys()
+		if err != nil {
+			return trace.WrapWithMessage(err, "unable to generate new keypairs")
+		}
+
+		params := RegisterParams{
+			Token:        onboarding.Token,
+			Servers:      []utils.NetAddr{*addr}, // TODO: multiple servers?
+			PrivateKey:   tlsPrivateKey,
+			PublicTLSKey: tlsPublicKey,
+			PublicSSHKey: sshPublicKey,
+			CipherSuites: utils.DefaultCipherSuites(),
+			CAPins:       onboarding.CAPins,
+			CAPath:       onboarding.CAPath,
+			Clock:        clockwork.NewRealClock(),
+		}
+
+		log.Info("attempting to generate new identity from token")
+		ident, err = newIdentityViaAuth(params)
+		if err != nil {
+			return trace.Wrap(err)
+		}
+
+		// Attach the ssh public key.
+		//ident.SSHPublicKeyBytes = sshPublicKey
+
+		// TODO: consider `memory` dest type for testing / ephemeral use / etc
+
+		log.Debug("attempting first connection using initial auth client")
+		authClient, err = authenticatedUserClientFromIdentity(ident, botConfig.AuthServer)
```

Comment:
this calls `auth.NewClient` which dials in the background, you may not see an error until the first API call. Could try Ping or something if you want to make sure the identity is good

Reply:
Oh, good catch! I'll add a ping test.